api/key/key.go

```diff
@@ -84,3 +92,34 @@ func Int(k string, v int) core.KeyValue {
 func Uint(k string, v uint) core.KeyValue {
 	return New(k).Uint(v)
 }
+
+// Infer creates a new key-value pair instance with a passed name and
+// automatic type inference. This is slower, and not type-safe.
+func Infer(k string, value interface{}) core.KeyValue {
+	switch v := value.(type) {
+	case bool:
+		return Bool(k, v)
+	case int:
+		return Int(k, v)
+	case uint:
+		return Uint(k, v)
+	case int32:
+		return Int32(k, v)
+	case int64:
+		return Int64(k, v)
+	case uint32:
+		return Uint32(k, v)
+	case uint64:
+		return Uint64(k, v)
+	case float32:
+		return Float32(k, v)
+	case float64:
+		return Float64(k, v)
+	case string:
+		return String(k, v)
+	case fmt.Stringer:
+		return Stringer(k, v)
```

Comment:
I wonder if you expect `Stringer` to be called directly by users. or if you would prefer to read `Infer()` calls. You could replace this line w/ `String(k, v.String())` and avoid exposing the function above. the performance is the same either way.

Reply:
I'd prefer to read Infer calls. especially if we wind up making this part of singular `SetAttribute` rather than `SetAttributes()`.

Reply:
I thought some more about this, and I wonder if we should use a method name like `Format()`.

I'm still in favor of `SetAttribute()`.

Reply:
> I thought some more about this, and I wonder if we should use a method name like `Format()`.
> 
> I'm still in favor of `SetAttribute()`.

Thoughts on doing SetAttribute using `infer` or `format`, a private method within the trace package? I dearly wish we could make the method not public, but instead an internal method callable by `api/trace` but not by other callers? raaaagh

Reply:
We can make an internal package, but I'm not sure exactly what you're proposing. Let's merge this and continue discussing?

Reply:
I'm good with merging this as is and doing a followup.